tests/subsys/audio_module/src/functional_test.c

```diff
@@ -0,0 +1,1801 @@
+	zassert_equal(handle.state, AUDIO_MODULE_STATE_STOPPED,
+		      "Stop state not AUDIO_MODULE_STATE_STOPPED (%d) rather %d",
+		      AUDIO_MODULE_STATE_STOPPED, handle.state);
+	zassert_mem_equal(&mod_context, handle_context, sizeof(struct mod_context), "Failed stop");
+}
+
+ZTEST(suite_audio_module_functional, test_start_fnct)
+{
+	int ret;
+
+	test_initialise_handle(&handle, &mod_description, &mod_context, NULL);
+	handle.state = AUDIO_MODULE_STATE_RUNNING;
+
+	ret = audio_module_start(&handle);
+	zassert_equal(ret, -EALREADY, "Start function did not return -EALREADY (%d): ret %d",
+		      -EALREADY, ret);
+	zassert_equal(handle.state, AUDIO_MODULE_STATE_RUNNING,
+		      "Start state not AUDIO_MODULE_STATE_RUNNING (%d) rather %d",
+		      AUDIO_MODULE_STATE_RUNNING, handle.state);
+
+	mod_context.test_string = NULL;
+	mod_context.test_uint32 = 0;
+	memset(&mod_context.config, 0, sizeof(struct mod_config));
+	test_initialise_handle(&handle, &mod_description, &mod_context, NULL);
+	handle_context = (struct mod_context *)handle.context;
+	handle.state = AUDIO_MODULE_STATE_STOPPED;
+
+	ret = audio_module_start(&handle);
+	zassert_equal(ret, 0, "Start function did not return successfully (0): ret %d", ret);
+	zassert_equal(handle.state, AUDIO_MODULE_STATE_RUNNING,
+		      "Start state not AUDIO_MODULE_STATE_RUNNING (%d) rather %d",
+		      AUDIO_MODULE_STATE_RUNNING, handle.state);
+	zassert_mem_equal(&mod_context, handle_context, sizeof(struct mod_context), "Failed start");
+
+	mod_context.test_string = NULL;
+	mod_context.test_uint32 = 0;
+	memset(&mod_context.config, 0, sizeof(struct mod_config));
+	test_initialise_handle(&handle, &mod_description, &mod_context, NULL);
+	handle_context = (struct mod_context *)handle.context;
+	handle.state = AUDIO_MODULE_STATE_CONFIGURED;
+
+	ret = audio_module_start(&handle);
+	zassert_equal(ret, 0, "Start function did not return successfully (0): ret %d", ret);
+	zassert_equal(handle.state, AUDIO_MODULE_STATE_RUNNING,
+		      "Start state not AUDIO_MODULE_STATE_RUNNING (%d) rather %d",
+		      AUDIO_MODULE_STATE_RUNNING, handle.state);
+	zassert_mem_equal(&mod_context, handle_context, sizeof(struct mod_context), "Failed start");
+}
+
+ZTEST(suite_audio_module_functional, test_disconnect_fnct)
+{
+	int ret;
+	int i, j, k;
+	int num_destinations;
+	char *test_inst_from_name = "TEST instance from";
+	char *test_inst_to_name = "TEST instance";
+	struct audio_module_handle handle_from;
+	struct audio_module_handle handles_to[TEST_CONNECTIONS_NUM];
+
+	for (k = 0; k < TEST_CONNECTIONS_NUM; k++) {
+		test_initialise_handle(&handles_to[k], &test_to_description, NULL, NULL);
+		snprintf(handles_to[k].name, CONFIG_AUDIO_MODULE_NAME_SIZE, "%s %d",
+			 test_inst_to_name, k);
+	}
+
+	test_from_description.type = AUDIO_MODULE_TYPE_INPUT;
+	test_to_description.type = AUDIO_MODULE_TYPE_OUTPUT;
+	for (i = AUDIO_MODULE_STATE_CONFIGURED; i <= AUDIO_MODULE_STATE_STOPPED; i++) {
```

Comment:
Let's talk about these from CONFIGURED to STOPPED. If someone changes these defines the for-loops will be affected.

Reply:
Yes, this is tricky and I would hope the test would fail if a new state was added.